runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/SamzaMetricsBundleProgressHandler.java

```diff
@@ -0,0 +1,150 @@
+
+  @Override
+  /**
+   * {@inheritDoc} Handles the bundle's completion report. Parses the monitoringInfos in the
+   * response, then updates the MetricsRegistry.
+   *
+   * <p>We attempt to construct a classic mode metricName
+   * ({transformUniqueName}:{className}:{metricName}). All the info should be in the labels, but we
+   * have fallbacks in case the labels don't exist.
+   *
+   * <p>Priorities for the transformUniqueName 1. Obtained transformUniqueName using the
+   * transformIdToUniqueName 2. The transformId provided by the monitoringInfo 3. The stepName
+   * provided by the runner, which maybe a result of fusing.
+   *
+   * <p>Priorities for the className 1. The namespace label 2. The monitoringInfo urn. Copying the
+   * implementation in MonitoringInfoMetricName.
+   *
+   * <p>Priorities for the metricName 1. The name label 2. The monitoringInfo urn. Copying the
+   * implementation in MonitoringInfoMetricName.
+   *
+   * @see
+   *     org.apache.beam.runners.core.metrics.MonitoringInfoMetricName#of(MetricsApi.MonitoringInfo)
+   */
+  public void onCompleted(BeamFnApi.ProcessBundleResponse response) {
+    for (MetricsApi.MonitoringInfo monitoringInfo : response.getMonitoringInfosList()) {
+      if (monitoringInfo.getPayload().isEmpty()) {
+        return;
+      }
+
+      String pTransformId =
+          monitoringInfo.getLabelsOrDefault(MonitoringInfoConstants.Labels.PTRANSFORM, stepName);
+      String transformUniqueName = transformIdToUniqueName.getOrDefault(pTransformId, pTransformId);
+      String className =
+          monitoringInfo.getLabelsOrDefault(
+              MonitoringInfoConstants.Labels.NAMESPACE, monitoringInfo.getUrn());
+      String userMetricName =
+          monitoringInfo.getLabelsOrDefault(
+              MonitoringInfoConstants.Labels.NAME, monitoringInfo.getLabelsMap().toString());
+
+      MetricsContainer metricsContainer = samzaMetricsContainer.getContainer(transformUniqueName);
+      MetricName metricName = MetricName.named(className, userMetricName);
+
+      switch (monitoringInfo.getType()) {
+        case SUM_INT64_TYPE:
+          Counter counter = metricsContainer.getCounter(metricName);
+          counter.inc(decodeInt64Counter(monitoringInfo.getPayload()));
+          break;
+
+        case DISTRIBUTION_INT64_TYPE:
+          Distribution distribution = metricsContainer.getDistribution(metricName);
+          DistributionData data = decodeInt64Distribution(monitoringInfo.getPayload());
+          distribution.update(data.sum(), data.count(), data.min(), data.max());
+          break;
+
+        case LATEST_INT64_TYPE:
+          Gauge gauge = metricsContainer.getGauge(metricName);
+          // Gauge doesn't expose update as public. This will reset the timestamp.
+
+          gauge.set(decodeInt64Gauge(monitoringInfo.getPayload()).value());
+          break;
+
+        default:
+          LOG.warn("Unsupported metric type {}", monitoringInfo.getType());
+      }
```

Comment:
this looks like a responsibility of `MetricsContainerImpl`. I'd move this within `SamzaMetricsContainer`

Reply:
`MonitoringInfo` is the data object used between the sdk worker and the runner; DistributionData/integer metric values are the data objected used within the runner.

I think it is the `BundleProgressHandler` to translate the external data type to internal data types, and pass it to the internal services.

Reply:
I don't agree. `MonitoringInfo` is an API defined as part of portability and isn't tied to worker/runner.
the object mutated is not an object belonging to this class (`BundleProgressHandler`). The handle is fetched from `MetricsContainer` which can change its own code and that percolates all the way to this class.

If you were to have a contract of `updateMetrics(monitoringInfo)` exposed by `MetricsContainer` this class doesn't have to deal with the evolution of this logic and stays core to its purpose of just signaling components to do the right thing instead of owning what that right thing is.

Reply:
Dropping the assumption of `MonitoringInfo` is the contract between worker/runner.

I'm okay with putting the code in `SamzaMetricsContainer`, but not `MetricsContainerImpl`.

Flink put some `updateMetrics(monitoringInfo)` code in `MetricsContainerImpl`, but as part of the translation, a MetricName has to be defined. The Flink formatting is different from Samza, making it not directly usable for us.

This is a violation of the dependency inversion principle, since it is no longer an abstraction, and doesn't work regardless who the client is. What would be better is `MetricsContainerImpl` provides methods usable by everyone, or just rename as `FlinkMetricsContainerImpl`.

After encountering this problem, I wanted to pull the translation to as close to the `BundleProgressHandler` as possible. Maybe I moved it too far up.

One concern for putting it in `SamzaMetricsContainer` though, is `SamzaMetricsContainer` serves both portability mode and classic mode. Why should we put a method only used by portable mode in the common space?

Reply:
portability or not, one can argue the responsibility is metrics handling and hence the logic should stay within SamzaMetricsContainer. Especially most of what the code does is pertained to metrics container and its internals.

IMO, the above code violates closed principle and leads to interactions with the component internals that `ProgressHandler` depends on it.

Reply:
"interactions with the component internals" is true, but it was a pattern already existing in `SamzaMetricsContainer`.

I can change it from 

```
          Distribution distribution = metricsContainer.getDistribution(metricName);
          DistributionData data = decodeInt64Distribution(monitoringInfo.getPayload());
          distribution.update(data.sum(), data.count(), data.min(), data.max());
```

to

```
          DistributionData data = decodeInt64Distribution(monitoringInfo.getPayload());
          Distribution distribution = metricsContainer.setDistribution(metricName, data);
```

but I think the parsing stays here.